pkg/cloudproviders/k8s/k8s_test.go

```diff
@@ -0,0 +1,283 @@
+	"github.com/spf13/viper"
+
+	"github.com/atlassian/gostatsd"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	core_v1 "k8s.io/api/core/v1"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	mainFake "k8s.io/client-go/kubernetes/fake"
+	kube_testing "k8s.io/client-go/testing"
+)
+
+const (
+	namespace = "foo"
+	podName1  = "pod1"
+	podName2  = "pod2"
+	nodeName  = "node1"
+
+	ipAddr  = "127.0.0.1"
+	ipAddr2 = "10.0.0.1"
+)
+
+var (
+	FullyQualifiedAnnotationWhitelistParam = ProviderName + "." + ParamAnnotationTagWhitelist
+	FullyQualifiedLabelWhitelistParam      = ProviderName + "." + ParamLabelTagWhitelist
+)
+
+type tagTestValue struct {
+	key         string // the key of the annotation/label
+	expectedKey string // the expected key of the tag
+	value       string // the value of the annotation/label & tag
+}
+
+var (
+	annotationTestValues = []tagTestValue{
+		{AnnotationPrefix + "tag1", "tag1", "value1"},
+		{"product.company.com/tag2", "tag2", "value2"},
+		{"tag3", "tag3", "value3"},
+	}
+	labelTestValues = []tagTestValue{
+		{"app", "app", "testApp"},
+	}
+)
+
+func pod() *core_v1.Pod {
+	annotations := map[string]string{}
+	for _, atv := range annotationTestValues {
+		annotations[atv.key] = atv.value
+	}
+	labels := map[string]string{}
+	for _, atv := range labelTestValues {
+		labels[atv.key] = atv.value
+	}
+
+	return &core_v1.Pod{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name:        podName1,
+			Namespace:   namespace,
+			Annotations: annotations,
+			Labels:      labels,
+		},
+		Status: core_v1.PodStatus{
+			PodIP: ipAddr,
+		},
+	}
+}
+
+func pod2() *core_v1.Pod {
+	// No labels/annotations for this pod
+	annotations := map[string]string{}
+	labels := map[string]string{}
+
+	return &core_v1.Pod{
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name:        podName2,
+			Namespace:   namespace,
+			Annotations: annotations,
+			Labels:      labels,
+		},
+		Status: core_v1.PodStatus{
+			PodIP: ipAddr2,
+		},
+	}
+}
+
+type testFixtures struct {
+	fakeClient    *mainFake.Clientset
+	cloudProvider gostatsd.CloudProvider
+	podsWatch     *watch.FakeWatcher
+}
+
+func setupTest(t *testing.T, test func(*testing.T, *testFixtures), v *viper.Viper, nn string) error {
+	fakeClient := mainFake.NewSimpleClientset()
+	podsWatch := watch.NewFake()
+	fakeClient.PrependWatchReactor("pods", kube_testing.DefaultWatchReactor(podsWatch, nil))
+
+	options := gostatsd.CloudProviderOptions{
+		Viper:    v,
+		Logger:   logrus.StandardLogger(),
+		Version:  "test",
+		NodeName: nn,
+	}
+	cloudProvider, err := NewProvider(options, fakeClient)
+	if err != nil {
+		return err
+	}
+
+	r, ok := cloudProvider.(gostatsd.Runner)
+	assert.True(t, ok)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go func() {
+		r.Run(ctx) // fork
+	}()
+
+	test(t, &testFixtures{
+		fakeClient:    fakeClient,
+		cloudProvider: cloudProvider,
+		podsWatch:     podsWatch,
+	})
+	return nil
```

Comment:
Could just call `require.*` and `t.FailNow()`/etc methods and then it would not need to return anything.

Reply:
I did this while I was working out how to wrap tests properly for table driven testing. With the current setup this does work much better, I'll change it. Changing it does complicate TestWatchNodeFailsNoNodeName a tiny bit though.